generate_grpc.sh

```diff
@@ -10,5 +10,17 @@ if [ ! -e ./proto/vendor/protobuf/src/google/protobuf ]; then
     wget https://raw.githubusercontent.com/protocolbuffers/protobuf/v3.9.0/src/google/protobuf/empty.proto -P $DIR
 fi
 
-# backing image manager
-protoc -I pkg/rpc/ -I proto/vendor/protobuf/src/ pkg/rpc/rpc.proto --go_out=plugins=grpc:pkg/rpc
+# Due to https://github.com/golang/protobuf/issues/1122, our .proto file names must be globally unique to avoid
+# namespace conflicts (https://protobuf.dev/reference/go/faq/#namespace-conflict). The best way to achieve this is to
+# include the whole Go import path in the name that gets compiled into the .pb.go file (e.g.
+# "github.com/longhorn/backing-image-manager/pkg/rpc.proto" instead of "rpc.proto"). This formulation does the
+# compilation inside a temporary directory prefixed with the full desired path before copying the .pb.go file out
+# (similar to the way https://github.com/container-storage-interface/spec/blob/master/lib/go/Makefile does it).
+PKG_DIR="pkg/rpc"
+TMP_DIR_BASE=".protobuild"
+TMP_DIR="${TMP_DIR_BASE}/github.com/longhorn/backing-image-manager/pkg/rpc"
+mkdir -p "${TMP_DIR}"
+cp "${PKG_DIR}/rpc.proto" "${TMP_DIR}/rpc.proto"
+protoc -I "${TMP_DIR_BASE}" -I "proto/vendor/protobuf/src/" "${TMP_DIR}/rpc.proto" --go_out=plugins=grpc:"${TMP_DIR_BASE}"
```

Comment:
Quick question. Is it ok to combine the import path like:
```
protoc -I "${TMP_DIR_BASE}" "proto/vendor/protobuf/src/" "${TMP_DIR}/rpc.proto" --go_out=plugins=grpc:"${TMP_DIR_BASE}"
```

Reply:
I spoke with @PhanLe1010 offline and we are agreed. I tried to find a good primary source for `protoc` (not `protoc-gen-go`) flags, but struggled. The `protoc --help` output provides the following direction:

```
  -IPATH, --proto_path=PATH   Specify the directory in which to search for
                              imports.  May be specified multiple times;
                              directories will be searched in order.  If not
                              given, the current working directory is used.
                              If not found in any of the these directories,
                              the --descriptor_set_in descriptors will be
                              checked for required proto file.
```